Firestore/Source/API/FIRDocumentSnapshot.m

```diff
@@ -76,6 +76,37 @@ - (instancetype)initWithFirestore:(FIRFirestore *)firestore
   return self;
 }
 
+// NSObject Methods
+- (BOOL)isEqual:(nullable id)other {
+  if (other == self) return YES;
+  // self class could be FIRDocumentSnapshot or subtype. So we compare with base type explicitly.
+  if (![other isKindOfClass:[FIRDocumentSnapshot class]]) return NO;
+
+  return [self isEqualToSnapshot:other];
+}
+
+- (BOOL)isEqualToSnapshot:(nullable FIRDocumentSnapshot *)snapshot {
+  if (self == snapshot) return YES;
+  if (snapshot == nil) return NO;
+  if (self.firestore != snapshot.firestore && ![self.firestore isEqual:snapshot.firestore])
+    return NO;
+  if (self.internalKey != snapshot.internalKey && ![self.internalKey isEqual:snapshot.internalKey])
+    return NO;
+  if (self.internalDocument != snapshot.internalDocument &&
+      (!self.internalDocument || ![self.internalDocument isEqual:snapshot.internalDocument]))
```

Comment:
You don't need the !self.internalDocument check do you? nil inequality will be caught by the first check as per your explanation above. :-)

Reply:
Yeah, the check is redundant. Removed.